File: cses/08_string_algorithms/1110_minimal-rotation/solution.py
import sys


def least_rotation(S):
    """Booth's algorithm: trả về chỉ số bắt đầu của rotation nhỏ nhất.

    S là xâu đã NHÂN ĐÔI (s + s), làm việc trên bytes (mỗi phần tử là int).
    """
    n2 = len(S)                 # = 2 * n
    f = [-1] * n2               # failure function (tương tự KMP)
    k = 0                       # vị trí bắt đầu rotation nhỏ nhất tìm được
    for j in range(1, n2):
        sj = S[j]
        i = f[j - k - 1]
        # So sánh với ứng viên bắt đầu từ k, dùng failure để nhảy lùi khi lệch;
        # gặp ký tự nhỏ hơn thì dời điểm bắt đầu k sang rotation tốt hơn.
        while i != -1 and sj != S[k + i + 1]:
            if sj < S[k + i + 1]:
                k = j - i - 1
            i = f[i]
        # Cập nhật failure function cho vị trí lệch so với điểm bắt đầu k.
        if sj != S[k + i + 1]:      # tức là i == -1
            if sj < S[k + i + 1]:   # khi i == -1 thì S[k+i+1] == S[k]
                k = j
            f[j - k] = -1
        else:
            f[j - k] = i + 1
    return k


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    s = data[0]                 # bytes
    n = len(s)
    doubled = s + s             # nhân đôi để tránh chỉ số modulo
    k = least_rotation(doubled)
    # Rotation nhỏ nhất là đoạn độ dài n bắt đầu tại k.
    out = doubled[k:k + n]
    sys.stdout.buffer.write(out)
    sys.stdout.buffer.write(b"\n")


if __name__ == "__main__":
    main()

File: cses/08_string_algorithms/1110_minimal-rotation/solution.cpp
#include <iostream>
#include <string>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string s;
    std::cin >> s;

    const int n = static_cast<int>(s.size());

    // Nhân đôi xâu để mọi rotation là một xâu con độ dài n bắt đầu tại 0..n-1,
    // tránh phải tính chỉ số modulo.
    const std::string doubled = s + s;

    // failure function (tương tự KMP); best là vị trí bắt đầu rotation nhỏ nhất.
    std::vector<int> failure(static_cast<std::size_t>(2 * n), -1);
    int best = 0;

    // Booth's algorithm: duyệt xâu nhân đôi, dùng failure để nhảy lùi khi lệch.
    for (int j = 1; j < 2 * n; ++j) {
        const char current = doubled[static_cast<std::size_t>(j)];
        int matched = failure[static_cast<std::size_t>(j - best - 1)];

        // So sánh ký tự hiện tại với ứng viên bắt đầu từ best; khi gặp ký tự
        // nhỏ hơn thì dời điểm bắt đầu best sang rotation tốt hơn.
        while (matched != -1 &&
               current != doubled[static_cast<std::size_t>(best + matched + 1)]) {
            if (current < doubled[static_cast<std::size_t>(best + matched + 1)]) {
                best = j - matched - 1;
            }
            matched = failure[static_cast<std::size_t>(matched)];
        }

        // Cập nhật failure function cho vị trí lệch so với điểm bắt đầu best.
        if (current != doubled[static_cast<std::size_t>(best + matched + 1)]) {
            if (current < doubled[static_cast<std::size_t>(best + matched + 1)]) {
                best = j;
            }
            failure[static_cast<std::size_t>(j - best)] = -1;
        } else {
            failure[static_cast<std::size_t>(j - best)] = matched + 1;
        }
    }

    // Rotation nhỏ nhất là đoạn độ dài n bắt đầu tại best.
    std::cout << doubled.substr(static_cast<std::size_t>(best),
                                static_cast<std::size_t>(n))
              << '\n';
    return 0;
}
